File: go/347.go
package main

import "container/heap"

type Heap [][]int

// Len implements heap.Interface
func (h *Heap) Len() int {
	return len(*h)
}

// Less implements heap.Interface
func (h *Heap) Less(i int, j int) bool {
	return (*h)[i][0] < (*h)[j][0]
}

// Swap implements heap.Interface
func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
	ret := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
  *h = append(*h, x.([]int))
}

func topKFrequent(nums []int, k int) []int {
	var h *Heap = &Heap{}
	heap.Init(h)
	var m map[int]int = make(map[int]int)
	// get count of num
	for _,v := range nums {
	  m[v]++
	}
	// import into heap, and controll size
	for i, v := range m {
	  heap.Push(h, []int{v, i})
	  if h.Len() > k {
	    heap.Pop(h)
	  }
	}
	var ret []int = make([]int, 0)
	for h.Len() > 0 {
	  ret = append(ret, heap.Pop(h).([]int)[1])
	}
	return ret
}

File: cpp/347.cpp
#include <functional>
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> topKFrequent(vector<int> &nums, int k) {
    priority_queue<pair<int, int>, vector<pair<int, int>>,
                   greater<pair<int, int>>>
        heap;
    unordered_map<int, 
      int> m;

    for (int n : nums) // to get m[num] = count
      m[n]++;

    // put data in map into heap to sort it and controll size as well
    for (auto p = m.begin(); p != m.end(); p++){
      heap.push({p->second, p->first});
      if (heap.size() > k)
        heap.pop();
    }

    // get ret
    vector<int> ret;
    while (heap.size() > 0) {
      ret.push_back(heap.top().second);
      heap.pop();
    }
    return ret;
  }
};
